File: programming_exercise_6/R/ex6_spam.R
# Copyright (C) 2014  Caleb Lo
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Machine Learning
# Programming Exercise 6: Spam Classification
# Problem: Use SVMs to determine whether various e-mails are spam (or non-spam)

# Load packages
library(e1071)
library(Rstem)

ReadKey <- function() {
  # Reads key press.
  #
  # Args:
  #   None.
  #
  # Returns:
  #   None.
  cat("Program paused. Press enter to continue.")
  line <- readline()
  return(0)
}

GetVocabList <- function() {
  # Returns array of words in a vocabulary list.
  #
  # Args:
  #   None.
  #
  # Returns:
  #   vocabList: Matrix that contains a vocabulary list.
  fullList <- as.matrix(scan("../vocab.txt", what=character()))
  numWords <- (dim(fullList)[1]) / 2
  if (numWords > 0) {
    vocabList <- matrix(0, numWords, 1)
    for (listIndex in 1:numWords) {
      vocabList[listIndex, ] <- fullList[2 * listIndex]
    }
  } else {
    stop('Insufficient number of words in vocabulary list')
  }
  return(vocabList)
}

ProcessEmail <- function(fileContents) {
  # Preprocesses e-mail and returns list of word indices.
  #
  # Args:
  #   fileContents: Vector of characters from an e-mail in a text file.
  #
  # Returns:
  #   wordIndices: Vector of indices of processed words in a vocabulary list.
  numChars <- dim(fileContents)[1]
  if (numChars > 0) {
    lowerFilCont <- matrix(0, numChars, 1)
    for (listIndex in 1:numChars) {
      lowerFilCont[listIndex, ] <- tolower(fileContents[listIndex, ])
      if (lowerFilCont[listIndex, ] == '\n') {
        lowerFilCont[listIndex, ] <- ' '
      }
    }
    lowerFilContStr <- lowerFilCont[1]
    for (listIndex in 2:numChars) {
      lowerFilContStr <- paste(lowerFilContStr, lowerFilCont[listIndex], 
                               sep='')
    }
    modFileContents <- gsub("<[^<>]+>", " ", lowerFilContStr)
    modFileContents <- gsub("[0-9]+", "number", modFileContents)
    modFileContents <- gsub("(http|https)://[^ ]+", "httpaddr", 
                            modFileContents)
    modFileContents <- gsub("[^ ]+@[^ ]+", "emailaddr", modFileContents)
    modFileContents <- gsub("[$]+", "dollar", modFileContents)
    print(sprintf("==== Processed Email ===="))

    # Remove punctuation in processed e-mail
    modFileContents <- gsub("[[:punct:]]", "", modFileContents)
    wordsModFileContents <- strsplit(modFileContents, " ")
    vocabList <- GetVocabList()
    wordIndices <- list()
    for (wordIndex in 1:length(wordsModFileContents[[1]])) {
      if (wordsModFileContents[[1]][wordIndex] != "") {
        currWord <- gsub("[^a-zA-Z0-9]", "", 
                         wordsModFileContents[[1]][wordIndex])

        # Apply porterStemmer in Rstem package
        stemCurrWord <- wordStem(currWord)
        if (nchar(stemCurrWord) >= 1) {

          # Search through vocabList for stemmed word
          for (vocabListIndex in 1:length(vocabList)) {
            if (stemCurrWord == vocabList[vocabListIndex]) {
              wordIndices <- rbind(wordIndices, vocabListIndex)
            }
          }

          # Display stemmed word
          print(sprintf("%s", stemCurrWord))
        }
      }
    }
    print(sprintf("========================="))
  } else {
    stop('Insufficient number of characters in e-mail')
  }
  return(wordIndices)
}

EmailFeatures <- function(emailWordIndices) {
  # Processes list of word indices and returns feature vector.
  #
  # Args:
  #   emailWordIndices: Vector of indices of processed words (from an e-mail) 
  #                     in a vocabulary list.
  #
  # Returns:
  #   featuresVec: Vector of booleans where the i-th entry indicates whether
  #                the i-th word in the vocabulary list occurs in the e-mail of
  #                interest.
  kNumDictWords <- 1899
  featuresVec <- matrix(0, kNumDictWords, 1)
  numEmailWordIndices <- length(emailWordIndices)
  if (numEmailWordIndices > 0) {
    for (wordIndex in 1:length(emailWordIndices)) {
      featuresVec[as.numeric(emailWordIndices[wordIndex])] <- 1
    }
  } else {
    stop('Insufficient number of processed word indices')
  }
  return(featuresVec)
}

# Use setwd() to set working directory to directory that contains this source 
# file
# Load file into R
print(sprintf("Preprocessing sample email (emailSample1.txt)"))
emailSample1 <- file("../emailSample1.txt", "rb")
fileContents <- readChar(emailSample1, 1)
for (cIndex in 2:file.info("../emailSample1.txt")$size) {
  fileContents <- rbind(fileContents, readChar(emailSample1, 1))
}
close(emailSample1)

# Extract features from file
emailWordIndices <- ProcessEmail(fileContents)
print(sprintf("Word Indices: "))
for (wordIndex in 1:length(emailWordIndices)) {
  print(sprintf("%d", as.numeric(emailWordIndices[wordIndex])))
}
returnCode <- ReadKey()
print(sprintf("Extracting features from sample email (emailSample1.txt)"))
emailSample1 <- file("../emailSample1.txt", "rb")
fileContents <- readChar(emailSample1, 1)
for (cIndex in 2:file.info("../emailSample1.txt")$size) {
  fileContents <- rbind(fileContents, readChar(emailSample1, 1))
}
close(emailSample1)
emailWordIndices <- ProcessEmail(fileContents)
featuresVec <- EmailFeatures(emailWordIndices)
print(sprintf("Length of feature vector: %d", length(featuresVec)))
print(sprintf("Number of non-zero entries: %d", sum(featuresVec > 0)))
returnCode <- ReadKey()

# Train a linear SVM for spam classification
spamTrain <- read.csv("../spamTrain.txt", header=FALSE)
xMat <- as.matrix(subset(spamTrain, select=-c(V1900)))
yVec <- as.vector(subset(spamTrain, select=c(V1900)))
print(sprintf("Training Linear SVM (Spam Classification)"))
print(sprintf("(this may take 1 to 2 minutes) ..."))
svmModel <- svm(xMat, y=yVec, scale=FALSE, type="C-classification", 
                kernel="linear", cost=0.1)
svmPred <- predict(svmModel, xMat)
print(sprintf("Training Accuracy: %.6f", 
              100 * apply(((as.numeric(svmPred) - 1) == yVec), 2, mean)))

# Test this linear spam classifier
spamTest <- read.csv("../spamTest.txt", header=FALSE)
xTestMat <- as.matrix(subset(spamTest, select=-c(V1900)))
yTestVec <- as.vector(subset(spamTest, select=c(V1900)))
print(sprintf("Evaluating the trained Linear SVM on a test set ..."))
testPred <- predict(svmModel, xTestMat)
print(sprintf("Test Accuracy: %.6f", 
              100 * apply(((as.numeric(testPred) - 1) == yTestVec), 2, mean)))
returnCode <- ReadKey()

# Determine the top predictors of spam
svmModelWeights <- t(svmModel$coefs) %*% (svmModel$SV)
svmModelWeightsSorted <- sort(svmModelWeights, decreasing=TRUE, 
                              index.return=TRUE)
vocabList <- GetVocabList()
print(sprintf("Top predictors of spam:"))
for (predIndex in 1:15) {
  print(sprintf("%s (%.6f)", vocabList[svmModelWeightsSorted$ix[predIndex]], 
                svmModelWeightsSorted$x[predIndex]))
}
returnCode <- ReadKey()

# Test this linear spam classifier on another e-mail
spamSample1 <- file("../spamSample1.txt", "rb")
fileContents <- readChar(spamSample1, 1)
for (cIndex in 2:file.info("../spamSample1.txt")$size) {
  fileContents <- rbind(fileContents, readChar(spamSample1, 1))
}
close(spamSample1)
emailWordIndices <- ProcessEmail(fileContents)
featuresVec <- EmailFeatures(emailWordIndices)
svmPred <- predict(svmModel, t(featuresVec))
print(sprintf("Processed spamSample1.txt"))
print(sprintf("Spam Classification: %d", as.numeric(svmPred) - 1))
print(sprintf("(1 indicates spam, 0 indicates not spam)"))

File: programming_exercise_6/Python/ex6_spam.py
# Copyright (C) 2015  Caleb Lo
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Machine Learning
# Programming Exercise 6: Spam Classification
# Problem: Use SVMs to determine whether various e-mails are spam (or non-spam)
from sklearn import svm
import numpy
import re
import Stemmer
import string


class Error(Exception):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return repr(self.value)


def get_vocab_list():
    """ Returns array of words in a vocabulary list.

    Args:
      None.

    Returns:
      vocab_list: Matrix that contains a vocabulary list.
    """
    vocab_file = open('../vocab.txt', 'r')
    vocab_list = []
    token_count = 0
    for line in vocab_file:
        for token in line.split():
            if (numpy.mod(token_count, 2) == 1):
                vocab_list.append(token)
            token_count = token_count + 1
    vocab_file.close()
    return vocab_list


def process_email(file_contents):
    """ Preprocesses e-mail and returns list of word indices.

    Args:
      file_contents: Vector of characters from an e-mail in a text file.
   
    Returns:
      word_indices: Vector of indices of processed words in a vocabulary list.

    Raises:
      An error occurs if the number of characters from the e-mail is 0.
      An error occurs if the number of processed words from the vocabulary list
      is 0.
    """
    num_email_chars = len(file_contents)
    if (num_email_chars == 0): raise Error('num_email_chars == 0')
    vocab_list = get_vocab_list()
    num_vocab_list_index = len(vocab_list)
    if (num_vocab_list_index == 0): raise Error('num_vocab_list_index == 0')
    lower_fil_cont = []
    for list_index in range(0, num_email_chars):
        lower_fil_cont.append(file_contents[list_index].lower())
        if (lower_fil_cont[list_index] == '\n'):
            lower_fil_cont[list_index] = ' '
    lower_fil_cont_str = lower_fil_cont[0]
    for list_index in range(0, len(lower_fil_cont)):
        lower_fil_cont_str = lower_fil_cont_str + lower_fil_cont[list_index]
    mod_file_contents = re.sub('<[^<>]+>', ' ', lower_fil_cont_str)
    mod_file_contents = re.sub('[0-9]+', 'number', mod_file_contents)
    mod_file_contents = re.sub('(http|https)://[^ ]+', 'httpaddr',
                               mod_file_contents)
    mod_file_contents = re.sub('[^ ]+@[^ ]+', 'emailaddr', mod_file_contents)
    mod_file_contents = re.sub('[$]+', 'dollar', mod_file_contents)
    print("==== Processed Email ====")

    # Remove punctuation in processed e-mail
    exclude = set(string.punctuation)
    mod_file_contents = (
        ''.join(ch for ch in mod_file_contents if ch not in exclude))
    words_mod_file_contents = mod_file_contents.split()
    word_indices = []

    # Apply porterStemmer in PyStemmer package
    stemmer = Stemmer.Stemmer('porter')
    for word_index in range(0, len(words_mod_file_contents)):
        curr_word = re.sub('[^a-zA-Z0-9]', '',
                           words_mod_file_contents[word_index])
        stem_curr_word = stemmer.stemWord(curr_word)
        if (len(stem_curr_word) >= 1) :

            # Search through vocab_list for stemmed word
            for vocab_list_index in range(0, num_vocab_list_index):
                if (stem_curr_word == vocab_list[vocab_list_index]):
                    word_indices.append(vocab_list_index)

            # Display stemmed word
            print("%s" % stem_curr_word)
    print("=========================")
    return word_indices


def email_features(email_word_indices):
    """ Processes list of word indices and returns feature vector.

    Args:
      email_word_indices: Vector of indices of processed words (from an e-mail) 
                          in a vocabulary list.

    Returns:
      features_vec: Vector of booleans where the i-th entry indicates whether
                    the i-th word in the vocabulary list occurs in the e-mail of
                    interest.

    Raises:
      An error occurs if the number of word indices is 0.
    """
    num_word_indices = len(email_word_indices)
    if (num_word_indices == 0): raise Error('num_word_indices == 0')
    num_dict_words = 1899
    features_vec = numpy.zeros((num_dict_words, 1))
    for word_index in range(0, num_word_indices):
        features_vec[email_word_indices[word_index]] = 1
    return features_vec


def main():
    """ Main function

    Raises:
      An error occurs if the number of processed words from an e-mail is 0.
      An error occurs if the number of training examples is 0.
      An error occurs if the number of test examples is 0.
    """
    print("Preprocessing sample email (emailSample1.txt)")
    email_sample_1 = open('../emailSample1.txt', 'r')
    file_contents = []
    while True:
        c = email_sample_1.read(1)
        if not c:
            break
        else:
            file_contents.append(c)
    email_sample_1.close()

    # Extract features from file
    email_word_indices = process_email(file_contents)
    num_word_indices = len(email_word_indices)
    if (num_word_indices == 0): raise Error('num_word_indices == 0')
    print("Word Indices: ")
    for word_index in range(0, num_word_indices):
        print("%d" % email_word_indices[word_index])
    input("Program paused. Press enter to continue.")
    print("")
    print("Extracting features from sample email (emailSample1.txt)")
    email_sample_1 = open('../emailSample1.txt', 'r')
    file_contents = []
    while True:
        c = email_sample_1.read(1)
        if not c:
            break
        else:
            file_contents.append(c)
    email_sample_1.close()
    email_word_indices = process_email(file_contents)
    features_vec = email_features(email_word_indices)
    print("Length of feature vector: %d" % features_vec.shape[0])
    print("Number of non-zero entries: %d" % numpy.sum(features_vec > 0))
    input("Program paused. Press enter to continue.")
    print("")

    # Train a linear SVM for spam classification
    spam_train_data = numpy.genfromtxt("../spamTrain.txt", delimiter=",")
    num_train_ex = spam_train_data.shape[0]
    if (num_train_ex == 0): raise Error('num_train_ex == 0')
    num_features = spam_train_data.shape[1]-1
    x_mat = spam_train_data[:, 0:num_features]
    y_vec = spam_train_data[:, num_features]
    print("Training Linear SVM (Spam Classification)")
    print("(this may take 1 to 2 minutes) ...")
    svm_model = svm.SVC(C=0.1, kernel='linear')
    svm_model.fit(x_mat, y_vec)
    pred_vec = svm_model.predict(x_mat)
    num_pred_match = 0
    for ex_index in range(0, num_train_ex):
        if (pred_vec[ex_index] == y_vec[ex_index]):
            num_pred_match = num_pred_match + 1
    print("Training Accuracy: %.6f" % (100*num_pred_match/num_train_ex))

    # Test this linear spam classifier
    spam_test_data = numpy.genfromtxt("../spamTest.txt", delimiter=",")
    num_test_ex = spam_test_data.shape[0]
    if (num_test_ex == 0): raise Error('num_test_ex == 0')
    x_test_mat = spam_test_data[:, 0:num_features]
    y_test_vec = spam_test_data[:, num_features]
    print("Evaluating the trained Linear SVM on a test set ...")
    pred_vec = svm_model.predict(x_test_mat)
    num_pred_match = 0
    for ex_index in range(0, num_test_ex):
        if (pred_vec[ex_index] == y_test_vec[ex_index]):
            num_pred_match = num_pred_match + 1
    print("Test Accuracy: %.6f" % (100*num_pred_match/num_test_ex))
    input("Program paused. Press enter to continue.")
    print("")

    # Determine the top predictors of spam
    svm_model_weights_sorted = numpy.sort(svm_model.coef_, axis=None)[::-1]
    svm_model_wts_sorted_idx = numpy.argsort(svm_model.coef_, axis=None)[::-1]
    vocab_list = get_vocab_list()
    print("Top predictors of spam:")
    for pred_index in range(0, 15):
        print("%s (%.6f)" % (vocab_list[svm_model_wts_sorted_idx[pred_index]],
                             svm_model_weights_sorted[pred_index]))
    input("Program paused. Press enter to continue.")
    print("")

    # Test this linear spam classifier on another e-mail
    spam_sample_1 = open('../spamSample1.txt', 'r')
    file_contents = []
    while True:
        c = spam_sample_1.read(1)
        if not c:
            break
        else:
            file_contents.append(c)
    spam_sample_1.close()
    email_word_indices = process_email(file_contents)
    features_vec = email_features(email_word_indices)
    pred_vec = svm_model.predict(numpy.transpose(features_vec))
    print("Processed spamSample1.txt")
    print("Spam Classification: %d" % pred_vec[0])
    print("(1 indicates spam, 0 indicates not spam)")

# Call main function
if __name__ == "__main__":
    main()
